pkg/controller/snapshot_controller.go

```diff
@@ -0,0 +1,773 @@
+	newSnapshot, err := ctrl.updateSnapshotStatus(snapshot, csiSnapshotStatus, time.Unix(0, timestamp), false)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create VolumeSnapshotContent in the database
+	volumeRef, err := ref.GetReference(scheme.Scheme, volume)
+
+	snapshotContent := &crdv1.VolumeSnapshotContent{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: contentName,
+		},
+		Spec: crdv1.VolumeSnapshotContentSpec{
+			VolumeSnapshotRef: &v1.ObjectReference{
+				Kind:       "VolumeSnapshot",
+				Namespace:  snapshot.Namespace,
+				Name:       snapshot.Name,
+				UID:        snapshot.UID,
+				APIVersion: "v1alpha1",
+			},
+			PersistentVolumeRef: volumeRef,
+			VolumeSnapshotSource: crdv1.VolumeSnapshotSource{
+				CSI: &crdv1.CSIVolumeSnapshotSource{
+					Driver:         driverName,
+					SnapshotHandle: snapshotID,
+					CreatedAt:      timestamp,
+				},
+			},
+		},
+	}
+
+	// Try to create the VolumeSnapshotContent object several times
+	for i := 0; i < ctrl.createSnapshotContentRetryCount; i++ {
+		glog.V(4).Infof("createSnapshot [%s]: trying to save volume snapshot data %s", snapshotKey(snapshot), snapshotContent.Name)
+		if _, err = ctrl.clientset.VolumesnapshotV1alpha1().VolumeSnapshotContents().Create(snapshotContent); err == nil || apierrs.IsAlreadyExists(err) {
+			// Save succeeded.
+			if err != nil {
+				glog.V(3).Infof("volume snapshot data %q for snapshot %q already exists, reusing", snapshotContent.Name, snapshotKey(snapshot))
+				err = nil
+			} else {
+				glog.V(3).Infof("volume snapshot data %q for snapshot %q saved", snapshotContent.Name, snapshotKey(snapshot))
+			}
+			break
+		}
+		// Save failed, try again after a while.
+		glog.V(3).Infof("failed to save volume snapshot data %q for snapshot %q: %v", snapshotContent.Name, snapshotKey(snapshot), err)
+		time.Sleep(ctrl.createSnapshotContentInterval)
+	}
+
+	if err != nil {
+		// Save failed. Now we have a storage asset outside of Kubernetes,
+		// but we don't have appropriate volumesnapshotdata object for it.
+		// Emit some event here and try to delete the storage asset several
+		// times.
+		strerr := fmt.Sprintf("Error creating volume snapshot data object for snapshot %s: %v.", snapshotKey(snapshot), err)
+		glog.Error(strerr)
+		ctrl.eventRecorder.Event(newSnapshot, v1.EventTypeWarning, "CreateSnapshotContentFailed", strerr)
+		return nil, err
+	}
+
+	// save succeeded, bind and update status for snapshot.
+	result, err := ctrl.bindandUpdateVolumeSnapshot(snapshotContent, newSnapshot)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Delete a snapshot
+// 1. Find the SnapshotContent corresponding to Snapshot
+//   1a: Not found => finish (it's been deleted already)
+// 2. Ask the backend to remove the snapshot device
+// 3. Delete the SnapshotContent object
+// 4. Remove the Snapshot from vsStore
+// 5. Finish
+func (ctrl *csiSnapshotController) DeleteSnapshotContentOperation(content *crdv1.VolumeSnapshotContent) error {
```

Comment:
lowercase `d` in `DeleteSnapshotContentOperation`? This function is not used outside of this pkg and `createSnapshotOperation` is lowercase

Reply:
Done